packages/compile-solidity/src/compilerSupplier/index.ts

```diff
@@ -0,0 +1,117 @@
+import path from "path";
+import fs from "fs";
+import semver from "semver";
+
+import { Docker, Local, Native, VersionRange } from "./loadingStrategies";
+
+const defaultSolcVersion = "0.5.16";
+
+export class CompilerSupplier {
+  private events: any;
+  private version: string;
+  private docker: boolean;
+  private compilerRoots: string[];
+  private strategyOptions: Partial<{
+    version: string;
+    docker: boolean;
+    compilerRoots: string[];
+    events: any; // represents a @truffle/events instance, which lacks types
+    spawn: {
+      maxBuffer: number;
+    };
+  }>;
+
+  constructor({ events, solcConfig }) {
+    const { version, docker, compilerRoots, spawn } = solcConfig;
+    this.events = events;
+    this.version = version ? version : defaultSolcVersion;
+    this.docker = docker;
+    this.compilerRoots = compilerRoots;
+    this.strategyOptions = {};
+    if (version) this.strategyOptions.version = this.version;
+    if (docker) this.strategyOptions.docker = compilerRoots;
+    if (compilerRoots) this.strategyOptions.compilerRoots = compilerRoots;
+    if (events) this.strategyOptions.events = events;
+    if (spawn) this.strategyOptions.spawn = spawn;
+  }
+
+  async load() {
+    const userSpecification = this.version;
+
+    let strategy;
```

Comment:
@haltman-at writes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687275690) about code duplication, which would've been addressed systematically in #4225.

Reply:
I posit that this duplication is not bad enough to warrant addressing at this time.